node: AssertionError on package version although it is available

I tried generating the sources from the following [lock file](https://github.com/zulip/zulip-desktop/blob/4b3608fc1ee58793560135ba53a00301592d4877/package-lock.json).

I am getting the following assertion errors:

```
Found 1 lockfiles.
Reading packages from lockfiles...
1475 packages read.
Generating packages [1473/1475] ms @ 2.0.0Task exception was never retrieved                             
future: <Task finished coro=<GeneratorProgress._generate() done, defined at ./node/flatpak-node-generator.py:1431> exception=AssertionError('entities versions available are 0.1.0, 0.1.1, 0.2.0, 0.2.1, 0.2.2, 0.3.0, 0.4.0, 0.4.1, 0.5.0, 1.0.0, 1.1.0, 1.1.1, 1.1.2, 2.0.0, 2.0.1, 2.0.2, 2.0.3, not https://registry.npmjs.org/entities/-/entities-1.1.2.tgz')>
Traceback (most recent call last):
  File "./node/flatpak-node-generator.py", line 1433, in _generate
    await self.module_provider.generate_package(package)
  File "./node/flatpak-node-generator.py", line 1110, in generate_package
    source = await self.resolve_source(package)
  File "./node/flatpak-node-generator.py", line 1086, in resolve_source
    f'{package.name} versions available are {", ".join(versions)}, not {package.version}'
AssertionError: entities versions available are 0.1.0, 0.1.1, 0.2.0, 0.2.1, 0.2.2, 0.3.0, 0.4.0, 0.4.1, 0.5.0, 1.0.0, 1.1.0, 1.1.1, 1.1.2, 2.0.0, 2.0.1, 2.0.2, 2.0.3, not https://registry.npmjs.org/entities/-/entities-1.1.2.tgz
Task exception was never retrieved
future: <Task finished coro=<GeneratorProgress._generate() done, defined at ./node/flatpak-node-generator.py:1431> exception=AssertionError('readable-stream versions available are 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.1.0, 0.2.0, 0.3.0, 0.3.1, 1.0.0, 1.0.1, 1.0.2, 1.0.15, 1.0.17, 1.1.7, 1.1.8, 1.1.9, 1.0.24, 1.0.25, 1.1.10, 1.0.25-1, 1.1.11, 1.0.26, 1.0.26-1, 1.1.11-1, 1.0.26-2, 1.1.12, 1.0.26-3, 1.0.26-4, 1.1.12-1, 1.0.27-1, 1.1.13-1, 1.0.31, 1.1.13, 1.0.32, 1.0.32-1, 1.0.33-1, 1.0.33-2, 1.0.33, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 1.0.34, 1.1.14, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.1.4, 2.1.5, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, 2.2.7, 2.2.8, 2.2.9, 2.2.10, 2.2.11, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.3.5, 2.3.6, 3.0.0-rc.1, 3.0.0-rc.2, 3.0.0-rc.3, 3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4, 3.0.5, 3.0.6, 3.1.0, 3.1.1, 3.2.0, 3.3.0, 3.4.0, 2.3.7, 3.5.0, 3.6.0, not https://registry.npmjs.org/readable-stream/-/readable-stream-3.6.0.tgz')>
Traceback (most recent call last):
  File "./node/flatpak-node-generator.py", line 1433, in _generate
    await self.module_provider.generate_package(package)
  File "./node/flatpak-node-generator.py", line 1110, in generate_package
    source = await self.resolve_source(package)
  File "./node/flatpak-node-generator.py", line 1086, in resolve_source
    f'{package.name} versions available are {", ".join(versions)}, not {package.version}'
AssertionError: readable-stream versions available are 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.1.0, 0.2.0, 0.3.0, 0.3.1, 1.0.0, 1.0.1, 1.0.2, 1.0.15, 1.0.17, 1.1.7, 1.1.8, 1.1.9, 1.0.24, 1.0.25, 1.1.10, 1.0.25-1, 1.1.11, 1.0.26, 1.0.26-1, 1.1.11-1, 1.0.26-2, 1.1.12, 1.0.26-3, 1.0.26-4, 1.1.12-1, 1.0.27-1, 1.1.13-1, 1.0.31, 1.1.13, 1.0.32, 1.0.32-1, 1.0.33-1, 1.0.33-2, 1.0.33, 2.0.0, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 1.0.34, 1.1.14, 2.1.0, 2.1.1, 2.1.2, 2.1.3, 2.1.4, 2.1.5, 2.2.0, 2.2.1, 2.2.2, 2.2.3, 2.2.4, 2.2.5, 2.2.6, 2.2.7, 2.2.8, 2.2.9, 2.2.10, 2.2.11, 2.3.0, 2.3.1, 2.3.2, 2.3.3, 2.3.4, 2.3.5, 2.3.6, 3.0.0-rc.1, 3.0.0-rc.2, 3.0.0-rc.3, 3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4, 3.0.5, 3.0.6, 3.1.0, 3.1.1, 3.2.0, 3.3.0, 3.4.0, 2.3.7, 3.5.0, 3.6.0, not https://registry.npmjs.org/readable-stream/-/readable-stream-3.6.0.tgz
```

All packages seem to have their expected version available for download. The error message suggest some parsing error in the version, as I don't expect to see the full URL in the `package.version` field.

Sorry to necrobump this, but I'm trying to build an app which lockfile has the same weird versioning scheme:

```json
"vue-loader-v16": {
  "version": "npm:vue-loader@16.1.2",
  "resolved": "https://registry.npmjs.org/vue-loader/-/vue-loader-16.1.2.tgz",
  "integrity": "sha512-8QTxh+Fd+HB6fiL52iEVLKqE9N1JSlMXLR92Ijm6g8PZrwIxckgpqjPDWRP5TWxdiPaHR+alUWsnu1ShQOwt+Q==",
  "dev": true,
  "optional": true,
  "requires": {
    "chalk": "^4.1.0",
    "hash-sum": "^2.0.0",
    "loader-utils": "^2.0.0"
  }
}
```

Here's the complete output when I run `./flatpak-node-generator.py npm -r ~/git/project/package-lock.json`:

```
Found 3 lockfiles.
Reading packages from lockfiles...
1793 packages read.
Generating packages [0/1793] /home/gui/git/flatpak-builder-tools/node/./flatpak-node-generator.py:1483: DeprecationWarning: The explicit passing of coroutine objects to asyncio.wait() is deprecated since Python 3.8, and scheduled for removal in Python 3.11.
  await asyncio.wait(map(self._generate, self.packages))
Generating packages [1792/1793] xtend @ 4.0.2Task exception was never retrieved                                                                                                               
future: <Task finished name='Task-1238' coro=<GeneratorProgress._generate() done, defined at /home/gui/git/flatpak-builder-tools/node/./flatpak-node-generator.py:1475> exception=AssertionError('vue-loader-v16 versions available are 16.0.0-beta.5, 16.0.0-beta.5.1, 16.0.0-beta.5.2, 16.0.0-beta.5.3, 16.0.0-beta.5.4, not npm:vue-loader@16.1.2')>
Traceback (most recent call last):
  File "/home/git/flatpak-builder-tools/node/./flatpak-node-generator.py", line 1477, in _generate
    await self.module_provider.generate_package(package)
  File "/home/gui/git/flatpak-builder-tools/node/./flatpak-node-generator.py", line 1141, in generate_package
    source = await self.resolve_source(package)
  File "/home/gui/git/flatpak-builder-tools/node/./flatpak-node-generator.py", line 1116, in resolve_source
    assert package.version in versions, \
AssertionError: vue-loader-v16 versions available are 16.0.0-beta.5, 16.0.0-beta.5.1, 16.0.0-beta.5.2, 16.0.0-beta.5.3, 16.0.0-beta.5.4, not npm:vue-loader@16.1.2

Wrote 2290 source(s).
```

--- a/node/flatpak_node_generator/package.py
+++ b/node/flatpak_node_generator/package.py
@@ -87,12 +87,12 @@ class PackageSource(abc.ABC):
 
 
 @dataclass(frozen=True, eq=True)
-class RegistrySource(PackageSource):
+class PackageFileSource(PackageSource):
     integrity: Optional[Integrity]
 
 
 @dataclass(frozen=True, eq=True)
-class ResolvedSource(RegistrySource):
+class PackageURLSource(PackageFileSource):
     resolved: str
 
     async def retrieve_integrity(self) -> Integrity:
@@ -103,6 +103,16 @@ class ResolvedSource(RegistrySource):
             assert url is not None, 'registry source has no resolved URL'
             metadata = await RemoteUrlMetadata.get(url, cachable=True)
             return metadata.integrity
+
+
+@dataclass(frozen=True, eq=True)
+class RegistrySource(PackageFileSource):
+    pass
+
+
+@dataclass(frozen=True, eq=True)
+class ResolvedSource(RegistrySource, PackageURLSource):
+    pass
 
 
 @dataclass(frozen=True, eq=True)

--- a/node/flatpak_node_generator/providers/npm.py
+++ b/node/flatpak_node_generator/providers/npm.py
@@ -27,6 +27,7 @@ from ..package import (
     LocalSource,
     Package,
     PackageSource,
+    PackageURLSource,
     RegistrySource,
     ResolvedSource,
 )
@@ -82,6 +83,8 @@ class NpmLockfileProvider(LockfileProvider):
                     source = ResolvedSource(
                         resolved=info['resolved'], integrity=integrity
                     )
+                elif version_url.scheme in {'http', 'https'}:
+                    source = PackageURLSource(resolved=version, integrity=integrity)
                 else:
                     source = RegistrySource(integrity=integrity)
 
@@ -280,6 +283,24 @@ class NpmModuleProvider(ModuleProvider):
             self.add_index_entry(source.resolved, metadata)
 
             await self.special_source_provider.generate_special_sources(package)
+
+        elif isinstance(source, PackageURLSource):
+            assert source.integrity is not None
+
+            self.gen.add_url_source(
+                url=source.resolved,
+                integrity=source.integrity,
+                destination=self.get_cacache_content_path(source.integrity),
+            )
+            self.add_index_entry(
+                url=source.resolved,
+                metadata=RemoteUrlMetadata(
+                    integrity=source.integrity,
+                    size=await RemoteUrlMetadata.get_size(
+                        source.resolved, cachable=True
+                    ),
+                ),
+            )
 
         # pyright: reportUnnecessaryIsInstance=false
         elif isinstance(source, GitSource):

--- a/node/flatpak_node_generator/providers/npm.py
+++ b/node/flatpak_node_generator/providers/npm.py
@@ -18,6 +18,7 @@ import re
 import shlex
 import textwrap
 import types
+import urllib.parse
 
 from ..integrity import Integrity
 from ..manifest import ManifestGenerator
@@ -59,6 +60,7 @@ class NpmLockfileProvider(LockfileProvider):
                 continue
 
             version: str = info['version']
+            version_url = urllib.parse.urlparse(version)
             alias_match = self._ALIAS_RE.match(version)
             if alias_match is not None:
                 name, version = alias_match.groups()
@@ -72,8 +74,8 @@ class NpmLockfileProvider(LockfileProvider):
                     from_ = from_[match.end('prefix') :]
 
                 source = self.parse_git_source(version, from_)
-            elif version.startswith('file:'):
-                source = LocalSource(path=version[len('file:') :])
+            elif version_url.scheme == 'file':
+                source = LocalSource(path=version_url.path)
             else:
                 integrity = Integrity.parse(info['integrity'])
                 if 'resolved' in info:

--- a/node/tests/data/packages/url-as-dep/package-lock.v1.json
+++ b/node/tests/data/packages/url-as-dep/package-lock.v1.json
@@ -1,0 +1,12 @@
+{
+  "name": "@flatpak-node-generator-tests/url-as-dep",
+  "version": "1.0.0",
+  "lockfileVersion": 1,
+  "requires": true,
+  "dependencies": {
+    "word-wrap": {
+      "version": "https://registry.npmjs.org/word-wrap/-/word-wrap-1.2.3.tgz",
+      "integrity": "sha512-Hz/mrNwitNRh/HUAtM/VT/5VH+ygD6DV7mYKZAtHOrbs8U7lvPS6xf7EJKMF0uW1KJCl0H701g3ZGus+muE5vQ=="
+    }
+  }
+}

--- a/node/tests/data/packages/url-as-dep/package-lock.v2.json
+++ b/node/tests/data/packages/url-as-dep/package-lock.v2.json
@@ -1,0 +1,30 @@
+{
+  "name": "@flatpak-node-generator-tests/url-as-dep",
+  "version": "1.0.0",
+  "lockfileVersion": 2,
+  "requires": true,
+  "packages": {
+    "": {
+      "name": "@flatpak-node-generator-tests/url-as-dep",
+      "version": "1.0.0",
+      "dependencies": {
+        "word-wrap": "https://registry.npmjs.org/word-wrap/-/word-wrap-1.2.3.tgz"
+      }
+    },
+    "node_modules/word-wrap": {
+      "version": "1.2.3",
+      "resolved": "https://registry.npmjs.org/word-wrap/-/word-wrap-1.2.3.tgz",
+      "integrity": "sha512-Hz/mrNwitNRh/HUAtM/VT/5VH+ygD6DV7mYKZAtHOrbs8U7lvPS6xf7EJKMF0uW1KJCl0H701g3ZGus+muE5vQ==",
+      "license": "MIT",
+      "engines": {
+        "node": ">=0.10.0"
+      }
+    }
+  },
+  "dependencies": {
+    "word-wrap": {
+      "version": "https://registry.npmjs.org/word-wrap/-/word-wrap-1.2.3.tgz",
+      "integrity": "sha512-Hz/mrNwitNRh/HUAtM/VT/5VH+ygD6DV7mYKZAtHOrbs8U7lvPS6xf7EJKMF0uW1KJCl0H701g3ZGus+muE5vQ=="
+    }
+  }
+}

--- a/node/tests/data/packages/url-as-dep/package.json
+++ b/node/tests/data/packages/url-as-dep/package.json
@@ -1,0 +1,7 @@
+{
+  "name": "@flatpak-node-generator-tests/url-as-dep",
+  "version": "1.0.0",
+  "dependencies": {
+    "word-wrap": "https://registry.npmjs.org/word-wrap/-/word-wrap-1.2.3.tgz"
+  }
+}

--- a/node/tests/test_providers.py
+++ b/node/tests/test_providers.py
@@ -117,6 +117,31 @@ async def test_missing_resolved_field(
     assert word_wrap_package_json.exists()
 
 
+async def test_url_as_dep(
+    flatpak_builder: FlatpakBuilder,
+    npm_provider_factory_spec: ProviderFactorySpec,
+    node_version: int,
+) -> None:
+    with ManifestGenerator() as gen:
+        await npm_provider_factory_spec.generate_modules(
+            'url-as-dep', gen, node_version
+        )
+
+    flatpak_builder.build(
+        sources=gen.ordered_sources(),
+        commands=[
+            npm_provider_factory_spec.install_command,
+            f"""node -e 'require("word-wrap")'""",
+        ],
+        use_node=node_version,
+    )
+
+    word_wrap_package_json = (
+        flatpak_builder.module_dir / 'node_modules' / 'word-wrap' / 'package.json'
+    )
+    assert word_wrap_package_json.exists()
+
+
 async def test_special_electron(
     flatpak_builder: FlatpakBuilder,
     provider_factory_spec: ProviderFactorySpec,

